File: 05_arrays/05_15_majority_element/majority_element.cpp
#include<iostream>
#include<vector>

using namespace std;

int majorityElement(vector<int> &nums) {
    int count = 1;
    int element = nums[0];

    for (int i = 1; i < count; i++)
    {
        if (count == 0)
        {
            count++;
            element = nums[i];
        }
        else if (element == nums[i])
        {
            count++;
        }
        else
        {
            count--;
        }
    }
    
    return element;
}

int main()
{
    vector<int> arr = {2, 2, 1, 1, 1, 2, 2};
    int n = arr.size();
    cout << "Before sorting:" << endl;
    for (int i = 0; i < n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
    int result = majorityElement(arr);
    cout << "The majority element is: " << result << endl;
    return 0;
}

File: 05_arrays/05_15_majority_element/majority_element.py
def majorityElement(nums: list):
    count, element = 1, nums[0]
    
    for i in range(1, len(nums)):
        if count == 0:
            count += 1
            element = nums[i]
        elif element == nums[i]:
            count += 1
        else:
            count -= 1
    
    return element

arr = [2, 2, 1, 1, 1, 2, 2]
print(*arr)
ans = majorityElement(arr)
print("The majority element is:", ans)